File: rust/overmax_app/src/play_state.rs
use crate::frame_utils::crop_roi;
use crate::frame_utils::region_mean_bgr;
use crate::ocr_engine::{OcrDetector, OcrTelemetry};
use crate::roi::RoiManager;
use crate::screen_capture::CapturedFrame;
use overmax_core::{GameSessionState, PlayContext};
use std::collections::VecDeque;

const BTN_MODE_MAX_DIST: f32 = 60.0;
const DIFF_MIN_BRIGHTNESS: f32 = 45.0;
const DIFF_CONFIDENT_MARGIN: f32 = 15.0;
const DIFFICULTIES: [&str; 4] = ["NM", "HD", "MX", "SC"];

type ButtonColorEntry = (&'static str, &'static [(u8, u8, u8)]);

#[derive(Clone, Debug, PartialEq)]
struct RawPlayState {
    context: Option<PlayContext>,
}

pub struct PlayStateDetector {
    history_size: usize,
    history: VecDeque<Option<RawPlayState>>,
    last_stable_state: Option<GameSessionState>,
    last_rate_thumb: Option<Vec<u8>>,
    last_rate_result: (Option<f32>, String, Option<OcrTelemetry>),
    last_rate_ocr_ts: f64,
}

impl PlayStateDetector {
    pub fn new(history_size: usize) -> Self {
        Self {
            history_size: history_size.max(1),
            history: VecDeque::new(),
            last_stable_state: None,
            last_rate_thumb: None,
            last_rate_result: (None, String::new(), None),
            last_rate_ocr_ts: 0.0,
        }
    }

    pub fn reset(&mut self) {
        self.history.clear();
        self.last_stable_state = None;
        self.last_rate_thumb = None;
        self.last_rate_result = (None, String::new(), None);
        self.last_rate_ocr_ts = 0.0;
    }

    pub fn detect(
        &mut self,
        frame: &CapturedFrame,
        rois: &RoiManager,
        song_id: Option<u32>,
        ocr: &OcrDetector,
        now: f64,
    ) -> (GameSessionState, Option<OcrTelemetry>) {
        let mode = detect_button_mode(frame, rois);
        let (diff, confident) = detect_difficulty(frame, rois);
        let is_max_combo = detect_max_combo(frame, rois);

        let mut telemetry = None;
        let context = if let (Some(sid), Some(m), Some(d)) = (song_id, mode, diff) {
            if confident {
                let mut rate = 0.0;
                if let Some(rate_roi) = rois.get_roi("rate") {
                    if let Some(rate_img) = crop_roi(frame, rate_roi) {
                        let thumb = crate::frame_utils::make_thumbnail(&rate_img);
                        let changed = if let Some(ref prev) = self.last_rate_thumb {
                            if let Some(ref t) = thumb {
                                crate::frame_utils::thumbnail_changed(
                                    t,
                                    Some(prev),
                                    2.0,
                                )
                            } else {
                                true
                            }
                        } else {
                            true
                        };

                        let should_ocr = changed
                            || (self.last_rate_result.0.is_none() && now - self.last_rate_ocr_ts >= 5.0);

                        if should_ocr {
                            if now - self.last_rate_ocr_ts >= 0.20 {
                                let res = ocr.detect_rate(&rate_img);
                                self.last_rate_result = res;
                                self.last_rate_thumb = thumb;
                                self.last_rate_ocr_ts = now;
                            }
                        }

                        rate = self.last_rate_result.0.unwrap_or(0.0);
                        telemetry = self.last_rate_result.2.clone();
                    }
                }

                Some(PlayContext {
                    song_id: sid,
                    mode: m,
                    diff: d,
                    rate,
                    is_max_combo: if rate > 0.0 { is_max_combo } else { false },
                })
            } else {
                None
            }
        } else {
            None
        };

        let raw = RawPlayState {
            context: context.clone(),
        };
        self.push_raw(raw);

        if let Some(stable) = self.stable_raw() {
            let state = GameSessionState {
                context: stable.context.clone(),
                is_stable: true,
            };
            self.last_stable_state = Some(state.clone());
            return (state, telemetry);
        }

        (GameSessionState {
            context,
            is_stable: false,
        }, telemetry)
    }

    fn push_raw(&mut self, raw: RawPlayState) {
        if self.history.len() == self.history_size {
            self.history.pop_front();
        }
        self.history.push_back(raw.context.is_some().then_some(raw));
    }

    fn stable_raw(&self) -> Option<&RawPlayState> {
        if self.history.len() != self.history_size {
            return None;
        }
        let first = self.history.front()?.as_ref()?;
        self.history
            .iter()
            .all(|item| item.as_ref() == Some(first))
            .then_some(first)
    }
}

pub fn detect_button_mode(frame: &CapturedFrame, rois: &RoiManager) -> Option<String> {
    let roi = rois.get_roi("btn_mode")?;
    let mean = region_mean_bgr(frame, roi);
    let mut best = (None, f32::INFINITY);
    for (mode, colors) in button_colors() {
        for color in colors {
            let dist = color_dist(mean, *color);
            if dist < best.1 {
                best = (Some(mode.to_string()), dist);
            }
        }
    }
    (best.1 <= BTN_MODE_MAX_DIST)
        .then_some(best.0)
        .flatten()
}

pub fn detect_difficulty(frame: &CapturedFrame, rois: &RoiManager) -> (Option<String>, bool) {
    let mut brightnesses = DIFFICULTIES
        .iter()
        .filter_map(|diff| {
            let roi = rois.get_diff_panel_roi(diff)?;
            let (b, g, r) = region_mean_bgr(frame, roi);
            Some((*diff, (f32::from(b) + f32::from(g) + f32::from(r)) / 3.0))
        })
        .collect::<Vec<_>>();
    brightnesses.sort_by(|a, b| b.1.total_cmp(&a.1));
    let Some((best, max_bright)) = brightnesses.first().copied() else {
        return (None, false);
    };
    if max_bright < DIFF_MIN_BRIGHTNESS {
        return (None, false);
    }
    let second = brightnesses.get(1).map_or(0.0, |item| item.1);
    (
        Some(best.to_string()),
        max_bright - second >= DIFF_CONFIDENT_MARGIN,
    )
}

pub fn detect_max_combo(frame: &CapturedFrame, rois: &RoiManager) -> bool {
    let Some(roi) = rois.get_roi("max_combo_badge") else {
        return false;
    };
    let (b, g, r) = region_mean_bgr(frame, roi);
    (f32::from(b) + f32::from(g) + f32::from(r)) / 3.0 >= 160.0
}

fn button_colors() -> [ButtonColorEntry; 4] {
    [
        ("4B", &[(0x55, 0x4F, 0x2D), (0x5A, 0x47, 0x0C)]),
        ("5B", &[(0xC6, 0xA9, 0x44)]),
        ("6B", &[(0x30, 0x94, 0xED)]),
        ("8B", &[(0x31, 0x14, 0x1D)]),
    ]
}

fn color_dist(left: (u8, u8, u8), right: (u8, u8, u8)) -> f32 {
    let db = f32::from(left.0) - f32::from(right.0);
    let dg = f32::from(left.1) - f32::from(right.1);
    let dr = f32::from(left.2) - f32::from(right.2);
    (db * db + dg * dg + dr * dr).sqrt()
}

#[cfg(test)]
mod tests {
    use super::{detect_button_mode, PlayStateDetector};
    use crate::roi::RoiManager;
    use crate::screen_capture::CapturedFrame;
    use overmax_core::SceneType;

    #[test]
    fn detects_button_mode_from_reference_color() {
        let mut frame = blank_frame();
        paint_rect(&mut frame, 80, 130, 85, 135, (0x55, 0x4F, 0x2D));
        let mut rois = RoiManager::new(1920, 1080);
        rois.set_scene(SceneType::Freestyle);
        assert_eq!(detect_button_mode(&frame, &rois), Some("4B".to_string()));
    }

    #[test]
    fn marks_state_stable_after_repeated_valid_frames() {
        let mut detector = PlayStateDetector::new(3);
        let mut frame = blank_frame();
        paint_rect(&mut frame, 80, 130, 85, 135, (0x55, 0x4F, 0x2D));
        paint_rect(&mut frame, 98, 488, 208, 516, (220, 220, 220));
        let mut rois = RoiManager::new(1920, 1080);
        rois.set_scene(SceneType::Freestyle);

        let ocr = crate::ocr_engine::OcrDetector::new();
        assert!(!detector.detect(&frame, &rois, Some(7), &ocr, 1.0).0.is_stable);
        assert!(!detector.detect(&frame, &rois, Some(7), &ocr, 2.0).0.is_stable);
        assert!(detector.detect(&frame, &rois, Some(7), &ocr, 3.0).0.is_stable);
    }

    fn blank_frame() -> CapturedFrame {
        CapturedFrame {
            width: 1920,
            height: 1080,
            bgra: vec![0; 1920 * 1080 * 4],
        }
    }

    fn paint_rect(
        frame: &mut CapturedFrame,
        x1: i32,
        y1: i32,
        x2: i32,
        y2: i32,
        bgr: (u8, u8, u8),
    ) {
        for y in y1..y2 {
            for x in x1..x2 {
                let idx = ((y * frame.width + x) * 4) as usize;
                frame.bgra[idx] = bgr.0;
                frame.bgra[idx + 1] = bgr.1;
                frame.bgra[idx + 2] = bgr.2;
            }
        }
    }
}

File: detection/play_state.py
"""
play_state.py - 게임 플레이 상태(모드, 난이도, 맥스콤보, 레이트) 통합 감지 모듈
"""

from __future__ import annotations
import math
from collections import deque
from typing import Optional, TYPE_CHECKING

import numpy as np

from constants import (
    BTN_COLORS,
    BTN_MODE_MAX_DIST,
    DIFF_MIN_BRIGHTNESS,
    DIFF_CONFIDENT_MARGIN,
)
from core.game_state import GameSessionState

if TYPE_CHECKING:
    from capture.roi_manager import ROIManager
    from detection.ocr import OcrDetector


# ------------------------------------------------------------------
# 내부 유틸 및 원시 감지 함수
# ------------------------------------------------------------------

def _color_dist(c1: tuple[int, int, int], c2: tuple[int, int, int]) -> float:
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(c1, c2)))


def _region_mean_bgr(
    frame: np.ndarray,
    roi: tuple[int, int, int, int],
) -> tuple[int, int, int]:
    h, w = frame.shape[:2]
    x1, y1, x2, y2 = roi
    x1, y1, x2, y2 = max(0, x1), max(0, y1), min(w, x2), min(h, y2)
    if x2 <= x1 or y2 <= y1:
        return (0, 0, 0)
    roi_img = frame[y1:y2, x1:x2]
    mean = roi_img.mean(axis=(0, 1))
    return int(mean[0]), int(mean[1]), int(mean[2])


def detect_button_mode(frame: np.ndarray, roiman: ROIManager) -> Optional[str]:
    roi = roiman.get_roi("btn_mode")
    mean_bgr = _region_mean_bgr(frame, roi)
    best_mode, best_dist = None, float("inf")
    for mode, colors in BTN_COLORS.items():
        for ref_bgr in colors:
            dist = _color_dist(mean_bgr, ref_bgr)
            if dist < best_dist:
                best_dist, best_mode = dist, mode
    return best_mode if best_dist <= BTN_MODE_MAX_DIST else None


def detect_difficulty(frame: np.ndarray, roiman: ROIManager) -> tuple[Optional[str], bool]:
    brightnesses = {
        diff: sum(_region_mean_bgr(frame, roiman.get_diff_panel_roi(diff))) / 3.0
        for diff in ["NM", "HD", "MX", "SC"]
    }
    sorted_diffs = sorted(brightnesses, key=lambda d: brightnesses[d], reverse=True)
    best_diff = sorted_diffs[0]
    max_bright = brightnesses[best_diff]
    if max_bright < DIFF_MIN_BRIGHTNESS:
        return None, False
    second_bright = brightnesses[sorted_diffs[1]] if len(sorted_diffs) > 1 else 0.0
    is_confident = (max_bright - second_bright) >= DIFF_CONFIDENT_MARGIN
    return best_diff, is_confident


def detect_max_combo(frame: np.ndarray, roiman: ROIManager) -> bool:
    roi = roiman.get_roi("max_combo_badge")
    b, g, r = _region_mean_bgr(frame, roi)
    return (r + g + b) / 3.0 >= 160


class PlayStateDetector:
    """
    모드, 난이도, 맥스 콤보, 그리고 달성률(Rate)을 통합적으로 감지합니다.
    히스테리시스를 통해 안정화된 상태를 판정하며, 상태가 확정되는 시점에 1회 Rate OCR을 수행합니다.
    """
    def __init__(self, ocr_detector: OcrDetector, history_size: int = 3):
        self.ocr_detector = ocr_detector
        self.history_size = max(1, history_size)
        
        self._history = deque(maxlen=self.history_size)
        self._last_stable_state: Optional[GameSessionState] = None
        self._ocr_done_for: Optional[tuple] = None # (song_id, mode, diff)

    def reset(self):
        """내부 히스토리 및 상태를 초기화합니다."""
        self._history.clear()
        self._last_stable_state = None
        self._ocr_done_for = None

    async def detect(
        self, 
        frame: np.ndarray, 
        roiman: ROIManager, 
        song_id: Optional[int],
        screen_mode: Optional[str] = None,
    ) -> GameSessionState:
        """
        현재 프레임을 분석하여 안정화된 게임 상태를 반환합니다.
        새로운 상태가 안정화되면 내부적으로 Rate OCR을 수행합니다.
        """
        raw_mode = detect_button_mode(frame, roiman)
        raw_diff, raw_confident = detect_difficulty(frame, roiman)
        raw_max_combo = detect_max_combo(frame, roiman)
        
        # 유효한 데이터인 경우만 히스토리에 추가
        current_raw = (song_id, raw_mode, raw_diff, raw_max_combo)
        is_valid_raw = all(v is not None for v in [song_id, raw_mode, raw_diff]) and raw_confident
        
        self._history.append(current_raw if is_valid_raw else None)
        
        # 안정화 여부 확인
        is_stable = False
        stable_res = None
        if len(self._history) == self.history_size and len(set(self._history)) == 1:
            if self._history[0] is not None:
                is_stable = True
                stable_res = self._history[0]
        
        if not is_stable:
            return GameSessionState(
                song_id=song_id,
                mode=raw_mode,
                diff=raw_diff,
                is_stable=False,
                is_max_combo=raw_max_combo,
                rate=None
            )

        # 안정화된 상태 처리
        s_sid, s_mode, s_diff, s_mc = stable_res
        
        # 새로운 상태인 경우 Rate OCR 시도
        current_state_key = (s_sid, s_mode, s_diff)
        rate = None
        
        if self._ocr_done_for == current_state_key:
            # 이미 이 상태에서 OCR을 했으면 기존 rate 유지 (있다면)
            if self._last_stable_state and self._last_stable_state.is_stable:
                rate = self._last_stable_state.rate
        else:
            # 새로운 상태 진입 -> OCR 수행
            rate_roi_name = "online_rate" if screen_mode == "ONLINE" else "rate"
            rate_roi = roiman.get_roi(rate_roi_name)
            # ROI 영역 크롭 (helpers 의 기능을 직접 사용하거나 roiman 확장 고려)
            h, w = frame.shape[:2]
            x1, y1, x2, y2 = rate_roi
            rate_img = frame[max(0, y1):min(h, y2), max(0, x1):min(w, x2)]
            
            rate, _ = await self.ocr_detector.detect_rate(rate_img)
            self._ocr_done_for = current_state_key

        new_state = GameSessionState(
            song_id=s_sid,
            mode=s_mode,
            diff=s_diff,
            is_stable=True,
            is_max_combo=s_mc,
            rate=rate
        )
        self._last_stable_state = new_state
        return new_state
